File: src/vscp/python/examples/event_handling.py
"""
Event Handling Example for VSCP Level 2 Client
Demonstrates event sending, receiving, and handling

Copyright (C) 2000-2026 Ake Hedman and contributors, the VSCP project <info@vscp.org>
MIT License
"""

import asyncio
import sys
import os

# Add the parent directory to the path so we can import vscp_level2
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from vscp_level2 import (
    VSCPLevel2Client,
    VSCPEvent,
    VSCPEventFilter,
    VSCPError,
    VSCPPriority,
    VSCPConstants
)


async def event_handling_example():
    """Demonstrate event handling with the VSCP Level 2 client"""
    print('=== VSCP Level 2 Client - Event Handling Example (Python) ===\n')
    
    try:
        # Create a new Level 2 client with debug enabled
        client = VSCPLevel2Client(
            interface='/dev/vscp0',
            flags=VSCPConstants.FLAG_ENABLE_DEBUG,
            response_timeout=3000
        )
        
        print('1. Created VSCP Level 2 client with debug enabled')
        
        # Initialize and connect
        await client.connect()
        print('2. Connected to interface')
        
        # Create and set a filter for measurement events
        filter_obj = VSCPEventFilter()
        filter_obj.set_class_and_type(10, 0)  # MEASUREMENT class, any type
        
        filter_result = client.set_filter(filter_obj)
        if filter_result == VSCPError.SUCCESS:
            print('3. Set event filter for measurement events')
        
        # Create different types of measurement events
        events = [
            # Temperature event
            VSCPEvent(
                vscp_class=10,
                vscp_type=6,
                guid=[0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
                      0x08, 0x09, 0x0A, 0x0B, 0x0C, 0x0D, 0x0E, 0x0F],
                data=[0x89, 0x02, 0x00, 0x19]  # 25.0°C
            ),
            
            # Humidity event
            VSCPEvent(
                vscp_class=10,
                vscp_type=7,
                guid=[0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
                      0x08, 0x09, 0x0A, 0x0B, 0x0C, 0x0D, 0x0E, 0x10],
                data=[0x89, 0x02, 0x00, 0x3C]  # 60.0% RH
            ),
            
            # Pressure event
            VSCPEvent(
                vscp_class=10,
                vscp_type=8,
                guid=[0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
                      0x08, 0x09, 0x0A, 0x0B, 0x0C, 0x0D, 0x0E, 0x11],
                data=[0x89, 0x02, 0x03, 0xE8]  # 1000 hPa
            )
        ]
        
        print('4. Created test measurement events')
        
        # Set timestamps and send events
        for i, event in enumerate(events):
            event.set_current_time()
            event.set_priority(VSCPPriority.NORMAL >> 5)
            
            print(f'\\n5.{i + 1}. Sending event {i + 1}/3:')
            print(f'     Class: {event.vscp_class}, Type: {event.vscp_type}')
            print(f'     Priority: {event.get_priority()}')
            print(f'     Hard coded: {event.is_hard_coded()}')
            print(f'     Data: {event.data}')
            
            send_result = await client.send(event)
            if send_result == VSCPError.SUCCESS:
                print('     ✅ Sent successfully')
            else:
                print(f'     ❌ Send failed: {send_result}')
            
            # Inject event into receive queue for demonstration
            client.inject_event(event)
            
            # Wait a bit between events
            await asyncio.sleep(0.5)
        
        print('\\n6. Attempting to receive events:')
        
        # Try to receive events
        for i in range(3):
            print(f'\\n6.{i + 1}. Receiving event {i + 1}/3:')
            
            receive_error, event = await client.receive(1000)
            if receive_error == VSCPError.SUCCESS and event:
                print(f'     ✅ Received event:')
                print(f'       Class: {event.vscp_class}, Type: {event.vscp_type}')
                print(f'       Timestamp: {event.year}-{event.month:02d}-{event.day:02d} {event.hour:02d}:{event.minute:02d}:{event.second:02d}')
                print(f'       Data: {event.data}')
                print(f'       GUID: {[f"{b:02x}" for b in event.guid]}')
            elif receive_error == VSCPError.TIMEOUT:
                print('     ⏰ Receive timeout')
            else:
                print(f'     ❌ Receive error: {receive_error}')
        
        # Get final statistics
        print('\\n7. Final statistics:')
        stats = client.get_statistics()
        print(f'   Events sent: {stats.events_sent}')
        print(f'   Events received: {stats.events_received}')
        print(f'   Errors: {stats.errors}')
        
        # Create an event with different serialization methods
        print('\\n8. Testing serialization methods:')
        json_event = VSCPEvent(
            vscp_class=10,
            vscp_type=6,
            data=[0x89, 0x02, 0x00, 0x20]  # 32.0°C
        )
        
        # Test dictionary conversion
        event_dict = json_event.to_dict()
        print(f'   Original event as dict: class={event_dict["vscp_class"]}, type={event_dict["vscp_type"]}')
        
        restored_from_dict = VSCPEvent.from_dict(event_dict)
        print(f'   Restored from dict: class={restored_from_dict.vscp_class}, type={restored_from_dict.vscp_type}')
        
        # Test JSON conversion
        event_json = json_event.to_json()
        print(f'   Event as JSON length: {len(event_json)} characters')
        
        restored_from_json = VSCPEvent.from_json(event_json)
        print(f'   Restored from JSON: class={restored_from_json.vscp_class}, type={restored_from_json.vscp_type}')
        print(f'   Restored data: {restored_from_json.data}')
        
        # Test priority and flags
        print('\\n9. Testing priority and flag methods:')
        test_event = VSCPEvent(vscp_class=10, vscp_type=6)
        
        priorities = [
            ('HIGH', VSCPPriority.HIGH),
            ('NORMAL', VSCPPriority.NORMAL), 
            ('MEDIUM', VSCPPriority.MEDIUM),
            ('LOW', VSCPPriority.LOW)
        ]
        
        for name, priority in priorities:
            test_event.set_priority(priority >> 5)
            print(f'   Priority {name}: {test_event.get_priority()} (head: 0x{test_event.head:04x})')
        
        # Test hard coded flag
        print(f'   Initially hard coded: {test_event.is_hard_coded()}')
        test_event.set_hard_coded(True)
        print(f'   After setting hard coded: {test_event.is_hard_coded()} (head: 0x{test_event.head:04x})')
        test_event.set_hard_coded(False)
        print(f'   After clearing hard coded: {test_event.is_hard_coded()} (head: 0x{test_event.head:04x})')
        
        # Disconnect
        await client.disconnect()
        
        print('\\n=== Event Handling Example Completed Successfully ===')
        
    except Exception as error:
        print(f'Error in event handling example: {error}')
        import traceback
        traceback.print_exc()
        return 1
    
    return 0


if __name__ == '__main__':
    exit_code = asyncio.run(event_handling_example())
    sys.exit(exit_code)

File: src/vscp/nodejs/examples/event-handling.js
/**
 * Event Handling Example for VSCP Level 2 Client
 * Demonstrates event sending, receiving, and handling
 * 
 * Copyright (C) 2000-2026 Ake Hedman and contributors, the VSCP project <info@vscp.org>
 * MIT License
 */

'use strict';

const { VscpLevel2Client, VscpEvent, VscpEventFilter, VSCP_CONSTANTS } = require('../lib');

async function eventHandlingExample() {
    console.log('=== VSCP Level 2 Client - Event Handling Example ===\n');
    
    try {
        // Create a new Level 2 client with debug enabled
        const client = new VscpLevel2Client({
            interface: '/dev/vscp0',
            flags: VSCP_CONSTANTS.FLAG_ENABLE_DEBUG,
            responseTimeout: 3000
        });
        
        // Set up event listeners
        client.on('connected', () => {
            console.log('📡 Client connected');
        });
        
        client.on('disconnected', () => {
            console.log('📡 Client disconnected');
        });
        
        client.on('eventSent', (event) => {
            console.log(`📤 Event sent - Class: ${event.vscp_class}, Type: ${event.vscp_type}`);
        });
        
        client.on('eventReceived', (event) => {
            console.log(`📥 Event received - Class: ${event.vscp_class}, Type: ${event.vscp_type}`);
            console.log(`   Data: [${event.data.join(', ')}]`);
        });
        
        client.on('error', (error) => {
            console.error('❌ Client error:', error.message);
        });
        
        console.log('1. Set up event listeners');
        
        // Initialize and connect
        await client.connect();
        
        // Create and set a filter for measurement events
        const filter = new VscpEventFilter();
        filter.setClassAndType(10, 0); // MEASUREMENT class, any type
        
        const filterResult = client.setFilter(filter);
        if (filterResult === VSCP_CONSTANTS.ERROR_SUCCESS) {
            console.log('2. Set event filter for measurement events');
        }
        
        // Create different types of measurement events
        const events = [
            // Temperature event
            new VscpEvent({
                vscp_class: 10,
                vscp_type: 6,
                GUID: [0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
                       0x08, 0x09, 0x0A, 0x0B, 0x0C, 0x0D, 0x0E, 0x0F],
                data: [0x89, 0x02, 0x00, 0x19]  // 25.0°C
            }),
            
            // Humidity event
            new VscpEvent({
                vscp_class: 10,
                vscp_type: 7,
                GUID: [0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
                       0x08, 0x09, 0x0A, 0x0B, 0x0C, 0x0D, 0x0E, 0x10],
                data: [0x89, 0x02, 0x00, 0x3C]  // 60.0% RH
            }),
            
            // Pressure event
            new VscpEvent({
                vscp_class: 10,
                vscp_type: 8,
                GUID: [0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
                       0x08, 0x09, 0x0A, 0x0B, 0x0C, 0x0D, 0x0E, 0x11],
                data: [0x89, 0x02, 0x03, 0xE8]  // 1000 hPa
            })
        ];
        
        console.log('3. Created test measurement events');
        
        // Set timestamps and send events
        for (let i = 0; i < events.length; i++) {
            const event = events[i];
            event.setCurrentTime();
            event.setPriority(VSCP_CONSTANTS.PRIORITY_NORMAL);
            
            console.log(`\\n4.${i + 1}. Sending event ${i + 1}/3:`);
            console.log(`     Class: ${event.vscp_class}, Type: ${event.vscp_type}`);
            console.log(`     Priority: ${event.getPriority()}`);
            console.log(`     Hard coded: ${event.isHardCoded()}`);
            console.log(`     Data: [${event.data.join(', ')}]`);
            
            const sendResult = await client.send(event);
            if (sendResult === VSCP_CONSTANTS.ERROR_SUCCESS) {
                console.log('     ✅ Sent successfully');
            } else {
                console.log(`     ❌ Send failed: ${sendResult}`);
            }
            
            // Inject event into receive queue for demonstration
            client.injectEvent(event);
            
            // Wait a bit between events
            await new Promise(resolve => setTimeout(resolve, 500));
        }
        
        console.log('\\n5. Attempting to receive events:');
        
        // Try to receive events
        for (let i = 0; i < 3; i++) {
            console.log(`\\n5.${i + 1}. Receiving event ${i + 1}/3:`);
            
            const receiveResponse = await client.receive(1000);
            if (receiveResponse.error === VSCP_CONSTANTS.ERROR_SUCCESS && receiveResponse.event) {
                const event = receiveResponse.event;
                console.log(`     ✅ Received event:`);
                console.log(`       Class: ${event.vscp_class}, Type: ${event.vscp_type}`);
                console.log(`       Timestamp: ${event.year}-${String(event.month).padStart(2, '0')}-${String(event.day).padStart(2, '0')} ${String(event.hour).padStart(2, '0')}:${String(event.minute).padStart(2, '0')}:${String(event.second).padStart(2, '0')}`);
                console.log(`       Data: [${event.data.join(', ')}]`);
                console.log(`       GUID: [${event.GUID.map(b => b.toString(16).padStart(2, '0')).join(', ')}]`);
            } else if (receiveResponse.error === VSCP_CONSTANTS.ERROR_TIMEOUT) {
                console.log('     ⏰ Receive timeout');
            } else {
                console.log(`     ❌ Receive error: ${receiveResponse.error}`);
            }
        }
        
        // Get final statistics
        console.log('\\n6. Final statistics:');
        const stats = client.getStatistics();
        console.log(`   Events sent: ${stats.eventsSent}`);
        console.log(`   Events received: ${stats.eventsReceived}`);
        console.log(`   Errors: ${stats.errors}`);
        
        // Create an event with JSON serialization
        console.log('\\n7. Testing JSON serialization:');
        const jsonEvent = new VscpEvent({
            vscp_class: 10,
            vscp_type: 6,
            data: [0x89, 0x02, 0x00, 0x20]  // 32.0°C
        });
        
        const eventJson = jsonEvent.toJSON();
        console.log('   Original event as JSON:', JSON.stringify(eventJson, null, 2));
        
        const restoredEvent = VscpEvent.fromJSON(eventJson);
        console.log('   Restored event class:', restoredEvent.vscp_class);
        console.log('   Restored event type:', restoredEvent.vscp_type);
        console.log('   Restored event data:', restoredEvent.data);
        
        // Disconnect
        await client.disconnect();
        
        console.log('\\n=== Event Handling Example Completed Successfully ===');
        
    } catch (error) {
        console.error('Error in event handling example:', error.message);
        console.error(error.stack);
        process.exit(1);
    }
}

// Run the example
if (require.main === module) {
    eventHandlingExample().catch(console.error);
}

module.exports = eventHandlingExample;